src/main/java/teammates/ui/webapi/BinCourseAction.java

```diff
@@ -22,18 +23,32 @@ void checkSpecificAccessControl() throws UnauthorizedAccessException {
         }
 
         String idOfCourseToBin = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
-        gateKeeper.verifyAccessible(logic.getInstructorForGoogleId(idOfCourseToBin, userInfo.id),
-                logic.getCourse(idOfCourseToBin), Const.InstructorPermissions.CAN_MODIFY_COURSE);
+
+        if (!isCourseMigrated(idOfCourseToBin)) {
+            CourseAttributes courseAttributes = logic.getCourse(idOfCourseToBin);
+            gateKeeper.verifyAccessible(logic.getInstructorForGoogleId(idOfCourseToBin, userInfo.id),
+                    courseAttributes, Const.InstructorPermissions.CAN_MODIFY_COURSE);
+            return;
+        }
+
+        Course course = sqlLogic.getCourse(idOfCourseToBin);
+        gateKeeper.verifyAccessible(sqlLogic.getInstructorByGoogleId(idOfCourseToBin, userInfo.id),
+                course, Const.InstructorPermissions.CAN_MODIFY_COURSE);
     }
 
     @Override
     public JsonResult execute() {
         String idOfCourseToBin = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
         try {
-            CourseAttributes courseAttributes = logic.getCourse(idOfCourseToBin);
-            courseAttributes.setDeletedAt(logic.moveCourseToRecycleBin(idOfCourseToBin));
-
-            return new JsonResult(new CourseData(courseAttributes));
+            if (!isCourseMigrated(idOfCourseToBin)) {
+                CourseAttributes courseAttributes = logic.getCourse(idOfCourseToBin);
+                courseAttributes.setDeletedAt(logic.moveCourseToRecycleBin(idOfCourseToBin));
+                return new JsonResult(new CourseData(courseAttributes));
+            }
+
+            Course course = sqlLogic.getCourse(idOfCourseToBin);
+            course.setDeletedAt(sqlLogic.moveCourseToRecycleBin(idOfCourseToBin));
```

Comment:
Currently the action layer returns a CourseData with the course fetched (in line 49). If nothing is returned from moveCourseToRecyclingBin, the deletedAt of the course being returned is not updated (as moveCourseToRecyclingBin fetches the course again in logic layer and sets that deletedAt field). 

Would it be better to fetch the course entity at the logic layer and set that deletedAt field, and then return the updated course?

Reply:
I mean since the course object is in managed state, when you set the deleteAt in `CoursesLogic`, the one in `BinCourseAction` should be updated as well. But yeah, what you suggested seems to be better!